### PR TITLE
More tests for variations on datatype declarations

### PR DESCRIPTION
**Pull request by rokopt**
Add tests for:

- Recursive pattern-matches
- Function types in datatype constructors
- Parameterized datatypes

### CODE REVIEW

File: test/examples/to-fix/circuit/Datatypes.ju [hunk 1]
```diff
@@ -25,6 +25,19 @@ sig patternMatch : Bar -> field
...
+-- ( probably a bug in the elaborator from frontend )"
+type AtomExp a = Atom a
+
+sig testAExp : AtomExp field
+let testAExp = Atom 1
+
+-- Fails to typecheck:
+-- "Error on toHR: Name "SExp" :| [] not in scope
+-- ( probably a bug in the elaborator from frontend )"
+type SExp a = Atom a | SPair (SExp a) (SExp a)
```
Comment (mariari) [inline test/examples/to-fix/circuit/Datatypes.ju:39]: This brings up an intersting test case we should have.

Namely in a working example can we break it with

```haskell
type atom-exp a = Atom a

type more-fun-exp = Atom Int.t
```

I think this is an error waiting to happen

File: test/examples/to-fix/llvm/Datatypes.ju [hunk 2]
```diff
@@ -18,6 +18,37 @@ sig patternMatch : Bar -> int
 let patternMatch (P x y) = x + y
 let patternMatch (Q x y z) = x + y + z
 
+type FunctionTypeInConstructor :
```
Comment (mariari) [inline test/examples/to-fix/llvm/Datatypes.ju:21]: Nit : For consistency can we use lower cased type names?

File: test/examples/to-fix/michelson/Datatypes.ju [hunk 3]
```diff
@@ -44,4 +75,32 @@ let main =
...
 
+  -- Typechecks, but fails to compile:
+  --  juvix: src/Juvix/Backends/Michelson/Compilation.hs:53:13-75:
+  --  Non-exhaustive patterns in Ann _ (Pi argUsage _ _) (LamM _ [name] body)
+  let ftict = fticTestTerm in
+
+  -- Typechecks, but fails to compile:
+  --  juvix: src/Juvix/Backends/Michelson/Compilation.hs:53:13-75:
+  --  Non-exhaustive patterns in Ann _ (Pi argUsage _ _) (LamM _ [name] body)
+  let sexpTest = ssize s1 in
```
Comment (mariari) [inline test/examples/to-fix/michelson/Datatypes.ju:86]: Yeah, currently we only compile Paired functions for Michelson, as those are the valid contract types

**Review by mariari — COMMENTED** [review]
Seems like a few nice test cases with variations that make it break.

Would be good to make some smaller tests from the failures here, along with some slightly evil tests (not too evil yet).

Also as a small nit, I'd like it if we can keep type names lower case.

File: test/examples/to-fix/circuit/Datatypes.ju (hunk 1 above)
Comment (rokopt) [reply test/examples/to-fix/circuit/Datatypes.ju]: Good idea, I added that one.

File: test/examples/to-fix/llvm/Datatypes.ju (hunk 2 above)
Comment (rokopt) [reply test/examples/to-fix/llvm/Datatypes.ju]: Sure, I've made the type names lowercase now.

**Comment by rokopt** [comment]
> Seems like a few nice test cases with variations that make it break.
> 
> Would be good to make some smaller tests from the failures here, along with some slightly evil tests (not too evil yet).

I'll definitely accumulate more over the course of #910 (and beyond!).

**Review by mariari — APPROVED** [review]
Seems good

**Pull request merged**